[InstCombine] Unpack scalable struct loads/stores.

davemgreen opened the pull request:
This teaches unpackLoadToAggregate and unpackStoreToAggregate to unpack scalable structs to individual loads/stores with insertvalues / extractvalues. The gep used for the offsets uses an i8 ptradd as opposed to a struct gep, as the struct geps are not supported and we canonicalize scalable geps to i8.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineLoadStoreAlloca.cpp:
```diff
@@ -722,11 +718,19 @@ static Instruction *unpackLoadToAggregate(InstCombinerImpl &IC, LoadInst &LI) {
+                            Addr,
+                            IC.Builder.CreateVScale(ConstantInt::get(
+                                DL.getIndexType(Addr->getType()),
+                                SL->getElementOffset(i).getKnownMinValue())),
```
Use `IRBuilder::CreateTypeSize()` and then you can use the same code for scalable and not?

nikic commented on llvm/test/Transforms/InstCombine/split_scalable_struct.ll:
```diff
@@ -0,0 +1,159 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -S -passes=instcombine < %s | FileCheck %s
+
+target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128-n8:16:32:64"
```
Is the data layout needed?

nikic reviewed (approved):
LGTM, thanks!

davemgreen committed:
[InstCombine] Unpack scalable struct loads/stores.

This teaches unpackLoadToAggregate and unpackStoreToAggregate to unpack
scalable structs to individual loads/stores with insertvalues / extractvalues.
The gep used for the offsets uses an i8 ptradd as opposed to a struct gep, as
the struct geps are not supported and we canonicalize scalable geps to i8.

davemgreen commented:
Thanks
